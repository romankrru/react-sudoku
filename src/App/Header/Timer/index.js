import React, { Fragment } from "react";

import styles from "./index.module.css";

const formatTime = (seconds) => {
	const ss = seconds % 60;
	const mm = Math.floor(seconds / 60) % 60;
	const hh = Math.floor(seconds / (60 * 60));
	const toStr = (t) => t.toString().padStart(2, "0");

	return [hh, mm, ss].map(toStr);
};

const Timer = (props) => {
	const [hh, mm, ss] = formatTime(props.seconds);

	return (
		<div className={styles.timer}>
			<div className={styles.timerClock}>
				{hh !== "00" && (
					<Fragment>
						{hh}
						<span className={styles.timerLegend}>h</span>{" "}
					</Fragment>
				)}

				{mm !== "00" && (
					<Fragment>
						{mm}
						<span className={styles.timerLegend}>min</span>{" "}
					</Fragment>
				)}

				<Fragment>
					{ss}
					<span className={styles.timerLegend}>sec</span>
				</Fragment>
			</div>

			<div className={styles.timerControl}></div>
		</div>
	);
};

export default Timer;
