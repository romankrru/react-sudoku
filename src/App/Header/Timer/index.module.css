.timer {
    display: flex;
    align-items: center;
    padding-top: 10px;

    color: white;
    font-size: 24px;
    line-height: 28px;
}

.timerClock {
    margin-right: 6px;
}

.timerLegend {
    font-size: 14px;
}

.timerControl {
    position: relative;

    width: 40px;
    height: 40px;

    border: 2px solid rgba(255, 255, 255, 0.4);
    cursor: pointer;

    transition: background-color .2s ease;
    border-radius: 50%;
}

.timerControlPause::before,
.timerControlPause::after {
    content: "";

    position: absolute;

    height: 18px;
    width: 4px;

    background-color: white;
    border-radius: 1px;
}

.timerControl::before {
    top: 9px;
    left: 11px;
}

.timerControl::after {
    top: 9px;
    left: 21px;
}

.timerControlPlay {
    background: url('../assets/triangle.svg') 13px center no-repeat;
}

.timerControl:hover {
    background-color: rgba(255, 255, 255, 0.2);
}