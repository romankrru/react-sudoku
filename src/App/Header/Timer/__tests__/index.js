import React from "react";
import { render, screen } from "@testing-library/react";

import Timer from "../";

const createTextMatchIgnoringMarkup = (text) => (content, node) => {
	const hasText = (node) => node.textContent === text;
	const nodeHasText = hasText(node);

	const childrenDontHaveText = Array.from(node.children).every(
		(child) => !hasText(child)
	);

	return nodeHasText && childrenDontHaveText;
};

const renderTimer = () => {
	const defaultProps = {
		toggle: jest.fn,
		isRunning: false,
		seconds: 0,
	};

	const view = render(<Timer {...defaultProps} />);

	const rerender = (newProps) =>
		view.rerender(<Timer {...{ ...defaultProps, ...newProps }} />);

	return { view, rerender };
};

describe("<Timer />", () => {
	it("renders and shows time in correct format", () => {
		const { view, rerender } = renderTimer();

		expect(
			screen.getByText(createTextMatchIgnoringMarkup("00sec"))
		).toBeInTheDocument();

		// 12sec
		rerender({ seconds: 12 });

		expect(
			screen.getByText(createTextMatchIgnoringMarkup("12sec"))
		).toBeInTheDocument();

		// 05min 05sec
		rerender({ seconds: 305 });

		expect(
			screen.getByText(createTextMatchIgnoringMarkup("05min 05sec"))
		).toBeInTheDocument();

		// 06h 30min 00sec
		rerender({ seconds: 23400 });

		expect(
			screen.getByText(createTextMatchIgnoringMarkup("06h 30min 00sec"))
		).toBeInTheDocument();

		expect(view.container).toMatchSnapshot();
	});

	it("renders correct controls depending on timer state", () => {
		const { view, rerender } = renderTimer();

		// renders Play icon
		expect(
			view.container.querySelector(".timerControlPlay")
		).toBeInTheDocument();

		expect(
			view.container.querySelector(".timerControlPause")
		).not.toBeInTheDocument();

		expect(view.container).toMatchSnapshot();

		// renders Stop icon
		rerender({ isRunning: true });

		expect(
			view.container.querySelector(".timerControlPause")
		).toBeInTheDocument();

		expect(
			view.container.querySelector(".timerControlPlay")
		).not.toBeInTheDocument();

		expect(view.container).toMatchSnapshot();
	});
});
