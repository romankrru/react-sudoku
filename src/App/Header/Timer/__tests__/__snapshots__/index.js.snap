// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Timer /> renders and shows time in correct format 1`] = `
<div>
  <div
    class="timer"
  >
    <div
      class="timerClock"
    >
      06
      <span
        class="timerLegend"
      >
        h
      </span>
       
      30
      <span
        class="timerLegend"
      >
        min
      </span>
       
      00
      <span
        class="timerLegend"
      >
        sec
      </span>
    </div>
    <div
      class="timerControl timerControlPlay"
    />
  </div>
</div>
`;

exports[`<Timer /> renders correct controls depending on timer state 1`] = `
<div>
  <div
    class="timer"
  >
    <div
      class="timerClock"
    >
      00
      <span
        class="timerLegend"
      >
        sec
      </span>
    </div>
    <div
      class="timerControl timerControlPlay"
    />
  </div>
</div>
`;

exports[`<Timer /> renders correct controls depending on timer state 2`] = `
<div>
  <div
    class="timer"
  >
    <div
      class="timerClock"
    >
      00
      <span
        class="timerLegend"
      >
        sec
      </span>
    </div>
    <div
      class="timerControl timerControlPause"
    />
  </div>
</div>
`;
